compat: remove deprecated VirtualSize

inknos opened the pull request:
Since compat version 1.43 the VirtualSize field in the GET /images/{name}/json, GET /images/json, and
GET /system/df responses is deprecated and will no longer be included in API v1.44. Use the Size field instead, which contains the same information.

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
compat: deprecate VirtualSize in version 1.43 and remove it in 1.44
```

Luap99 commented:
What happens if a client talks in an older API version? Should there be some version check around this instead?

Luap99 commented:
I think the policy for compat is match whatever docker is doing. If you can run the latest docker and use API v1.40 (or anything below the removal) and it still returns that field then we need to do as well.

For libpod we have a strict semver commitment so any breaking changes such as remove response field should wait for the next major version.

inknos commented:
@Luap99 I implemented the api versioned deprecation, and added tests too. but now my eyes are crossing because tests are failing inconsistently.

- tests for /images/json fail for 1.44
- tests for /images/id/json pass
- tests for /system/df fail for 1.43

but the change should be the same, I am removing the same field conditionally for all the endpoints. am I missing something obvious?

Luap99 commented:
> @Luap99 I implemented the api versioned deprecation, and added tests too. but now my eyes are crossing because tests are failing inconsistently.
> 
>     * tests for /images/json fail for 1.44

Looking at the code this is a different endpoint which uses GetImage**s**() and that does not call imageDataToImageInspect at all. It seems like we are just returning the podman/libpod output .

You would need to modify it in the same way as https://github.com/containers/podman/commit/3ef33653ffdfa5a2cd22a3e7e3302fd8c73fd7bd there

> 
>     * tests for /images/id/json pass
> 
>     * tests for /system/df fail for 1.43

AFAIK the test only has a single image and as such Size == UniqueSize so Size - UniqueSize = 0 and as such because the type has omitempty it is not added to the json thus it always fails even on an older version.

It seems list/inspect on the other hand sets VirtualSize == Size so maybe df should have the same. I really have no idea what VirtualSize is meant represent exactly but I think df calculates it wrong then

> 
> 
> but the change should be the same, I am removing the same field conditionally for all the endpoints. am I missing something obvious?

Luap99 commented:
also you will need to add something like `//nolint:staticcheck // Deprecated field, but we must keep setting it for older API users` to the places VirtualSize is set as the linter complains otherswise

inknos commented on test/apiv2/45-system.at:
```diff
@@ -38,6 +38,29 @@ cid=$(jq -r '.Id' <<<"$output")
+EOF
+
+# Test VirtualSize field is present in API v1.43 for system/df (backward compatibility)
+size=$(t GET /v1.43/system/df .Images[0].Size)
+uniquesize=$(t GET /v1.43/system/df .Images[0].UniqueSize)
```
whops, it stayed from a squash I did with experimenting with the tests. it didn't work so I removed it but the lines stayed. I'll fix them

mheon commented:
/lgtm
